### ISSUE TITLE
Revise support for action methods in XtxtUML

### ISSUE THREAD

**Issue by djnemeth** [ELTE-Soft/txtUML#385]
The following table shows the action methods available in JtxtUML and their XtxtUML-specific alternatives. Please note that JtxtUML action methods can be used as conventional methods inside XtxtUML nevertheless — this issue is strictly about XtxtUML-specific _language elements_ to be provided instead of / along conventional API method calls.

| JtxtUML action method | XtxtUML-specific alternative |
| --- | --- |
| [`create(«type», «ctor args...»)`](https://github.com/ELTE-Soft/txtUML/blob/c94570311efb3ca1a1c8ee728cc0b6ef2daf912c/dev/plugins/hu.elte.txtuml.api.model/src/hu/elte/txtuml/api/model/Action.java#L53) | — |
| [`createWithName(«type», «name», «ctor args...»)`](https://github.com/ELTE-Soft/txtUML/blob/c94570311efb3ca1a1c8ee728cc0b6ef2daf912c/dev/plugins/hu.elte.txtuml.api.model/src/hu/elte/txtuml/api/model/Action.java#L83) | — |
| [`delete(«object»)`](https://github.com/ELTE-Soft/txtUML/blob/c94570311efb3ca1a1c8ee728cc0b6ef2daf912c/dev/plugins/hu.elte.txtuml.api.model/src/hu/elte/txtuml/api/model/Action.java#L103) | `delete «object»` |
| [`connect(«endA», «portA», «endB», «portB»)`](https://github.com/ELTE-Soft/txtUML/blob/c94570311efb3ca1a1c8ee728cc0b6ef2daf912c/dev/plugins/hu.elte.txtuml.api.model/src/hu/elte/txtuml/api/model/Action.java#L125) | — |
| [`connect(«parentPort», «childEnd», «childPort»)`](https://github.com/ELTE-Soft/txtUML/blob/c94570311efb3ca1a1c8ee728cc0b6ef2daf912c/dev/plugins/hu.elte.txtuml.api.model/src/hu/elte/txtuml/api/model/Action.java#L145) | — |
| [`link(«endA», «objA», «endB», «objB»)`](https://github.com/ELTE-Soft/txtUML/blob/c94570311efb3ca1a1c8ee728cc0b6ef2daf912c/dev/plugins/hu.elte.txtuml.api.model/src/hu/elte/txtuml/api/model/Action.java#L174) | — |
| [`unlink(«endA», «objA», «endB», «objB»)`](https://github.com/ELTE-Soft/txtUML/blob/c94570311efb3ca1a1c8ee728cc0b6ef2daf912c/dev/plugins/hu.elte.txtuml.api.model/src/hu/elte/txtuml/api/model/Action.java#L201) | — |
| [`start(«object»)`](https://github.com/ELTE-Soft/txtUML/blob/c94570311efb3ca1a1c8ee728cc0b6ef2daf912c/dev/plugins/hu.elte.txtuml.api.model/src/hu/elte/txtuml/api/model/Action.java#L217) | — |
| [`send(«signal», «reception»)`](https://github.com/ELTE-Soft/txtUML/blob/c94570311efb3ca1a1c8ee728cc0b6ef2daf912c/dev/plugins/hu.elte.txtuml.api.model/src/hu/elte/txtuml/api/model/Action.java#L241) | `send «signal» to «port»` |
| [`send(«signal», «object»)`](https://github.com/ELTE-Soft/txtUML/blob/c94570311efb3ca1a1c8ee728cc0b6ef2daf912c/dev/plugins/hu.elte.txtuml.api.model/src/hu/elte/txtuml/api/model/Action.java#L259) | `send «signal» to «object»` |
| [`log(«message»)`](https://github.com/ELTE-Soft/txtUML/blob/c94570311efb3ca1a1c8ee728cc0b6ef2daf912c/dev/plugins/hu.elte.txtuml.api.model/src/hu/elte/txtuml/api/model/Action.java#L269) | — |
| [`logError(«message»)`](https://github.com/ELTE-Soft/txtUML/blob/c94570311efb3ca1a1c8ee728cc0b6ef2daf912c/dev/plugins/hu.elte.txtuml.api.model/src/hu/elte/txtuml/api/model/Action.java#L279) | — |

We should consider the following:
- It can be seen that XtxtUML-specific action support is not extensive, to put it mildly.
- It's clearly inconsistent that we provide a keyword for `delete`, but not for `start`, `log` and `logError`.
- Introducing new language elements would extend our already long list of keywords, which could cause additional nuisance for our users.
- Getting rid of all action-related XtxtUML elements and relying only on JtxtUML-provided action methods would be an option as well.
- Being able to use port instances instead of receptions is a crucial advantage of the XtxtUML-specific `send` statement. See the following example:
  
  JtxtUML:
  
  ``` Java
  Action.send(new Tick(), port(OutTickPort.class).required::reception);
  ```
  
  XtxtUML:
  
  ``` Java
  send new Tick() to this->(OutTickPort);
  ```
  
  Therefore I don't recommend scrapping XtxtUML action elements in general.
- Even if we introduced keywords for all _syntactically simple_ JtxtUML actions (i.e. `start`, `delete`, `log` and `logError`), it would still be difficult to find suitable keyword-simplifications for _syntactically complex_ actions (i.e. `create`, `createWithName`, `connect`, `link` and `unlink`). Of course, we could make the choice that complex actions would be made available only through JtxtUML methods, as it is in the current implementation. (This seems a bit inconsistent for me, though.)

In summary, the questions are:
1. Should we provide XtxtUML-specific action elements _only_ for signal sending?
2. If no, which other action methods do you think we should provide language elements for? Do you have any suggestions regarding the syntax?

@ELTE-Soft/all, what's your opinion?

This depends on #390.

**Comment by djnemeth** [comment]
As far as I know, Java's `new` operator can be used instead of `Action.create`. Due to static validation, currently it seems to be safer than our action method. @kovacsgabor, I have distant memories about you saying that it's worth it nevertheless, but I don't remember why. Can you tell me what its advantages are?

**Comment by djnemeth** [comment]
According to @kovacsgabor, currently `Action.create` has no benefits over the `new` operator.

**Comment by djnemeth** [comment]
The team have decided that we should provide XtxtUML-specific alternatives only to those action methods, whose syntax is more complex (i.e. the ones with more than one argument).

According to this concept, we should
- drop the `delete` keyword,
- add support for linking/unlinking and connecting with the following syntax:
 <pre><b>link</b> objA <b>as</b> endA, objB <b>as</b> endB <b>via</b> MyAssoc;</pre>
 <pre><b>connect</b> portA <b>as</b> endA, portB <b>as</b> endB <b>via</b> MyConnector;</pre>
 where the *as* parts would be omittable, as long as the corresponding ends can be inferred.

**Comment by djnemeth** [comment]
Recently it has become apparent that the inherently inconsistent language induced by providing XtxtUML-specific alternatives only for a subset of the action methods would be less desirable than the one supporting all action methods at the cost of using more keywords. Therefore our previous decision has been modified as follows:
- keep the `delete` keyword,
- add support for `Action.start` and `Action.log` with keywords `start` and `log`, respectively,
- find a suitable syntax, and add support, for `Action.createWithName` and `Action.logError` (the former one will become especially important once multidebugging is realized, see #534).

**Comment by kovacsgabor** [comment]
The word `as` is already a keyword in XtxtUML, isn't it? Because if we are not introducing a new keyword with it then I would suggest for the X-variant of `Action.createWithName`:
 - `create «object» as «name»` or
 - `new «object» as «name»`

And for `Action.logError`, my first thought is `log-error` but it is not very nice.

**Comment by kovacsgabor** [comment]
And don't forget some recently added new `Action` methods!
 - `collect(«obj»...)` is a variadic function to create a new `Any` collection (unordered, non-unique, 0..*).
 - `collectIn(«collectionType», «obj»...)` is a variadic function to create a new collection of the given type.

They are named differently in Java to avoid conflicts but I think this differentiation is not necessary in X so the `collect` keyword could be applied for both.

**Comment by djnemeth** [comment]
It is tempting to choose `new «object» as «name»` for `Action.createWithName`, but in this case we would violate our policy of keeping the two language variants as close as possible. On the other hand, using `new` from Java is very convenient and intuitive.

I think the core of this problem is the unfortunate redundancy caused by `Action.create` and Java's `new` in JtxtUML. Dropping one for the other might work. Personally, I like `new` more than `create`, because the declarative nature of the former seems to fit better into contexts they're usually used in. Not to mention that disallowing `new` would likely be really annoying for users.

I understand that despite all this, the corresponding UML action is still called `CreateObjectAction`. The question is that between UML and Java, where do we want to position our language? Create instead of `new` might be too alienating. But what if we drop `Action.create`, rename `Action.createWithName` to `Action.newAs` or `Action.newWithName`, and in X we support `new` with an optional `as` clause only? At first glance, `Action.new...` looks strange, maybe right because the `Action` prefix implies something imperative.

Of course we could also keep `new`, `Action.create`, `Action.createWithName` in J and support `new`, `new ... as ...`, `create` and `create ... as ...` in X. Apart from the inherent redundancy and the additional inconsistency that `new ... as ...` would be translated to `Action.create...`, the contrast between `as` and `with name` is also confusing. Add to the mix that `as` is also used for casting in X.

Overall, this is a mess I strongly believe we should address. @kovacsgabor, what do you think?

**Comment by kovacsgabor** [comment]
(For simplicity, I will speak about the version of the JtxtUML API after the refactoring which is to be merged with #555. This pr does not change the syntax of the API methods, but modify their body at some points.)

The original thought behind the introduction and later the forcing of `Action.create` in J is to have a syntax for object creation which resembles the `Action.delete` call. The team thought that it is good for the language to have a similar syntax for object creation and deletion which can therefore be easily paired in one's mind. However, object deletion has never gained so much attention over the development of txtUML mainly because of Java being our host language and people in the team tending to think in Java rather than in UML.

Also, the use of an action language method enabled us in the past to do some checks before the object creation or add some other parameters which do not appear in the constructor of a model class. Currently (that is, after #555) the `Action.create` call has no parameters apart from the type of the to-be-created object and its constructor parameters and only checks whether it is run on a model executor thread.

Therefore, the simple `Action.create` could easily be replaced with `new` if the team decided so.

The case of `Action.createWithName` is also simpler than it seems because the name of an object can be modified even after its creation (see `ModelClasses` utility class in the `api.model.external` package - this is available in external libraries). So it can easily be replaced with an `Action.setName` method which simply sets the name after the creation of the model object.

It is also a possibility to create a `setName` and `getName` method in `ModelClass` but I would not recommend this because this name is only used by the txtUML model executor, it is not an actual attribute of the model classes (and users may wish to add a separate `name` attribute to their classes).

**Comment by kovacsgabor** [comment]
If the team decides to go with `new` instead of `Action.create` in J as well then I would suggest to consider `start`, `send` and `delete` as methods of the `ModelClass` type instead of the `Action` class. These are unlikely to appear as method names of model classes created in txtUML because of their use in the language itself and from a Java developer perspective, they might look better in many cases.

**Comment by djnemeth** [comment]
The team have decided on the following:
- keep `new ...`, `Action.create` and `Action.createWithName` in J,
- support `new ...`, `new ... as ...`, `create ...` and `create ... as ...` in X and map them to `new ...`, `Action.createWithName`, `Action.create` and `Action.createWithName` respectively.

**Comment by djnemeth** [comment]
Action methods related to the new collection framework will be implemented as part of #539.
